feat(wallet): propagate status with the pending tx status changed

stefandunca opened the pull request:
Replace usage of `eth_getTransactionByHash` with `eth_getTransactionReceipt`
when polling for changes. `eth_getTransactionReceipt` delivers also the
status of the transaction.

Update tests to account for the new changes
Propagate status
Refactor the `pendingtxtracker.go` file to emit notifications with a new payload structure that includes transaction identity and deletion status.
Includes: #4492

Closes status-desktop [#13124](https://github.com/status-im/status-desktop/issues/13124)

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (28)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | e26f00da | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/1/) | 2024-01-04 04:34:51 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/1//consoleText) |
| :heavy_check_mark: | e26f00da | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/1/) | 2024-01-04 04:36:09 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240104-043206-e26f00-pr4523.zip) |
| :heavy_check_mark: | e26f00da | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/1/) | 2024-01-04 04:36:59 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240104-043206-e26f00-pr4523.zip) |
| :heavy_check_mark: | e26f00da | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/1/) | 2024-01-04 04:38:04 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240104-043206-e26f00-pr4523.aar) |
| | | | | | | |
| :x: | 2ca78634 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/2/) | 2024-01-05 04:34:46 | ~23 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/2/consoleText) |
| :x: | 2ca78634 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/2/) | 2024-01-05 04:34:50 | ~23 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/2/consoleText) |
| :x: | 2ca78634 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/2/) | 2024-01-05 04:35:48 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/2/consoleText) |
| :heavy_multiplication_x: | 2ca78634 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/2/) | 2024-01-05 04:36:36 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/2//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | 1c6f710d | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/3/) | 2024-01-05 17:53:17 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/3//consoleText) |
| :heavy_check_mark: | 1c6f710d | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/3/) | 2024-01-05 17:54:16 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240105-175158-1c6f71-pr4523.aar) |
| :heavy_check_mark: | 1c6f710d | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/3/) | 2024-01-05 17:54:44 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240105-175203-1c6f71-pr4523.zip) |
| :heavy_check_mark: | 1c6f710d | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/3/) | 2024-01-05 17:55:33 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240105-175203-1c6f71-pr4523.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 6a7bf0be | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/4/) | 2024-01-06 04:11:01 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/4//consoleText) |
| :heavy_check_mark: | 6a7bf0be | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/4/) | 2024-01-06 04:11:36 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240106-040837-6a7bf0-pr4523.aar) |
| :heavy_check_mark: | 6a7bf0be | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/4/) | 2024-01-06 04:11:53 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240106-040840-6a7bf0-pr4523.zip) |
| :heavy_check_mark: | 6a7bf0be | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/4/) | 2024-01-06 04:12:01 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240106-040840-6a7bf0-pr4523.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 110640c4 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/5/) | 2024-01-06 04:21:02 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/5//consoleText) |
| :heavy_check_mark: | 110640c4 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/5/) | 2024-01-06 04:21:23 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240106-041951-110640-pr4523.zip) |
| :heavy_check_mark: | 110640c4 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/5/) | 2024-01-06 04:21:53 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240106-041947-110640-pr4523.aar) |
| :heavy_check_mark: | 110640c4 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/5/) | 2024-01-06 04:28:11 | ~8 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240106-041947-110640-pr4523.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 96426301 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/6/) | 2024-01-08 18:14:33 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/6//consoleText) |
| :heavy_check_mark: | 96426301 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/6/) | 2024-01-08 18:15:32 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240108-181305-964263-pr4523.aar) |
| :heavy_check_mark: | 96426301 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/6/) | 2024-01-08 18:16:31 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240108-181310-964263-pr4523.zip) |
| :heavy_check_mark: | 96426301 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/6/) | 2024-01-08 18:23:43 | ~10 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240108-181310-964263-pr4523.zip) |
| | | | | | | |
| :heavy_check_mark: | de535a90 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/7/) | 2024-01-08 18:33:36 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240108-183155-de535a-pr4523.zip) |
| :heavy_multiplication_x: | de535a90 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/7/) | 2024-01-08 18:34:38 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/7//consoleText) |
| :heavy_check_mark: | de535a90 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/7/) | 2024-01-08 18:37:59 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240108-183155-de535a-pr4523.aar) |
| :heavy_check_mark: | de535a90 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/7/) | 2024-01-08 18:39:25 | ~7 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240108-183155-de535a-pr4523.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 3807f18e | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/8/) | 2024-01-08 21:26:48 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240108-212454-3807f1-pr4523.zip) |
| :heavy_check_mark: | 3807f18e | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/8/) | 2024-01-08 21:27:11 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240108-212454-3807f1-pr4523.aar) |
| :heavy_check_mark: | 3807f18e | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/8/) | 2024-01-08 21:28:27 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240108-212454-3807f1-pr4523.zip) |
| :heavy_multiplication_x: | 3807f18e | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/8/) | 2024-01-08 21:45:15 | ~20 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/8//consoleText) |
| | | | | | | |
| :heavy_check_mark: | d501301a | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/9/) | 2024-01-09 00:08:35 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240109-000650-d50130-pr4523.zip) |
| :heavy_check_mark: | d501301a | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/9/) | 2024-01-09 00:09:14 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240109-000650-d50130-pr4523.aar) |
| :heavy_check_mark: | d501301a | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-4523/10/) | 2024-01-09 00:10:01 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240109-000838-d50130-pr4523.zip) |
| :heavy_check_mark: | d501301a | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/9/) | 2024-01-09 00:10:23 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240109-000650-d50130-pr4523.zip) |
| :heavy_check_mark: | d501301a | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/10/) | 2024-01-09 00:10:50 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240109-000917-d50130-pr4523.aar) |
| :heavy_check_mark: | d501301a | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-4523/10/) | 2024-01-09 00:13:25 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240109-001027-d50130-pr4523.zip) |
| :heavy_multiplication_x: | d501301a | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/9/) | 2024-01-09 00:44:21 | ~37 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/9//consoleText) |
| :heavy_check_mark: | d501301a | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/10/) | 2024-01-09 01:21:56 | ~37 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4523/10//consoleText) |

J-Son89 commented on transactions/pendingtxtracker.go:
```diff
@@ -24,12 +25,14 @@ import (
 
 const (
-	// EventPendingTransactionUpdate is emitted when a pending transaction is updated (added or deleted). Carries StatusChangedPayload in message
+	// EventPendingTransactionUpdate is emitted when a pending transaction is updated (added or deleted). Carries PendingTxUpdatePayload in message
```
curious, does this change break the current api? i.e you will have to make updates to desktop for this to work with the current feature?

stefandunca replied:
It doesn't. For now, we still use the deprecated `watch` API https://github.com/status-im/status-go/blob/405d468e0e74e78a89f0653ad6514c50af9d1d6f/services/wallet/api.go#L283

J-Son89 commented on transactions/pendingtxtracker.go:
```diff
@@ -139,6 +151,18 @@ func (tm *PendingTxTracker) fetchAndUpdateDB(ctx context.Context) bool {
+}
+
+func (nr *nullableReceipt) UnmarshalJSON(data []byte) error {
+	if string(data) == "null" {
```
not sure of the best practices etc in this codebase/Go in general but I generally prefer to use named variables and let the code speak then commenting etc. consider something like:
```
const transactionNotAvailable = string(data) == "null"
if  transactionNotAvailable {
	return nil
}
```

stefandunca replied:
Good idea, thanks for the heads up.

stefandunca committed:
chore(wallet): refactor test helpers and update event names

Refactor the test helpers for pending transactions by extracting the mock
implementations and helper functions into a new file `transactions/testhelpers.go`.

Update event names for clarity and consistency in `pendingtxtracker.go`.

Updates status-desktop: [#13124](https://github.com/status-im/status-desktop/issues/13124)

ghstack-source-id: d85756b80fefd3bf7590637cb7f4dc1af615cecd
Pull Request resolved: https://github.com/status-im/status-go/pull/4492

status-im-auto commented:
#### :heavy_check_mark: [status-go/prs/android/PR-4523#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-4523/7/) :small_blue_diamond: ~6 min 16 sec :small_blue_diamond: de535a90 :small_blue_diamond: [:package: android package](https://status-im-prs.ams3.digitaloceanspaces.com/status-go-android-240108-183155-de535a-pr4523.aar)

stefandunca committed:
feat(wallet): propagate status with the pending tx status changed

Replace usage of `eth_getTransactionByHash` with `eth_getTransactionReceipt`
when polling for changes. `eth_getTransactionReceipt` delivers also the
status of the transaction.

Update tests to account for the new changes
Propagate status with the update event

Refactoring of the `pendingtxtracker.go` file to emit notifications with
a new payload structure that includes transaction identity and deletion status.

Closes status-desktop [#12120](https://github.com/status-im/status-desktop/issues/13124)

ghstack-source-id: 936bff5a417cb7a4aa42c93fc9dacb03fc35943b
Pull Request resolved: https://github.com/status-im/status-go/pull/4523

dlipicar commented on transactions/testhelpers.go:
❤️
